store/tikv/tikvrpc/tikvrpc.go

```diff
@@ -354,3 +372,138 @@ func (resp *Response) GetRegionError() (*errorpb.Error, error) {
+	case CmdBatchRollback:
+		resp.BatchRollback, err = client.KvBatchRollback(ctx, req.BatchRollback)
+	case CmdScanLock:
+		resp.ScanLock, err = client.KvScanLock(ctx, req.ScanLock)
+	case CmdResolveLock:
+		resp.ResolveLock, err = client.KvResolveLock(ctx, req.ResolveLock)
+	case CmdGC:
+		resp.GC, err = client.KvGC(ctx, req.GC)
+	case CmdDeleteRange:
+		resp.DeleteRange, err = client.KvDeleteRange(ctx, req.DeleteRange)
+	case CmdRawGet:
+		resp.RawGet, err = client.RawGet(ctx, req.RawGet)
+	case CmdRawPut:
+		resp.RawPut, err = client.RawPut(ctx, req.RawPut)
+	case CmdRawDelete:
+		resp.RawDelete, err = client.RawDelete(ctx, req.RawDelete)
+	case CmdRawScan:
+		resp.RawScan, err = client.RawScan(ctx, req.RawScan)
+	case CmdCop:
+		resp.Cop, err = client.Coprocessor(ctx, req.Cop)
+	case CmdCopStream:
+		// Use context to support timeout for grpc streaming client.
+		ctx1, cancel := context.WithCancel(ctx)
+		var stream tikvpb.Tikv_CoprocessorStreamClient
+		stream, err = client.CoprocessorStream(ctx1, req.Cop)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		wrapStream := NewCopStreamResponse(stream, nil, cancel, ch)
+		// Read the first streaming response to get CopStreamResponse.
+		// This can make error handling much easier, because SendReq() retry on
+		// region error automatically.
+		var first *coprocessor.Response
+		first, err = wrapStream.Recv()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		wrapStream.Response = first
+		resp.CopStream = wrapStream
+	case CmdMvccGetByKey:
+		resp.MvccGetByKey, err = client.MvccGetByKey(ctx, req.MvccGetByKey)
+	case CmdMvccGetByStartTs:
+		resp.MvccGetByStartTS, err = client.MvccGetByStartTs(ctx, req.MvccGetByStartTs)
+	case CmdSplitRegion:
+		resp.SplitRegion, err = client.SplitRegion(ctx, req.SplitRegion)
+	default:
+		return nil, errors.Errorf("invalid request type: %v", req.Type)
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return resp, nil
+}
+
+// TimeoutItem is used to implement grpc stream timeout.
+type TimeoutItem struct {
+	cancel   context.CancelFunc
+	timeout  time.Time
+	complete int32 // atomic variable, 0 means false
+}
+
+// Recv overrides the stream client Recv() function.
+func (resp *CopStreamResponse) Recv() (*coprocessor.Response, error) {
+	item := &TimeoutItem{
+		cancel: resp.cancel,
+		// TODO: Make timeout a field in CopStreamResponse.
+		timeout: time.Now().Add(20 * time.Second),
+	}
+	resp.timeoutCh <- item
+
+	ret, err := resp.Tikv_CoprocessorStreamClient.Recv()
+
+	atomic.StoreInt32(&item.complete, 1)
+	return ret, errors.Trace(err)
+}
+
+// CheckStreamTimeoutLoop runs periodically to check is there any stream request timeouted.
+// TimeoutItem is an object to track stream requests, call this function with "go CheckStreamTimeoutLoop()"
+func CheckStreamTimeoutLoop(ch <-chan *TimeoutItem) {
+	ticker := time.NewTicker(200 * time.Millisecond)
+	defer ticker.Stop()
+	array := make([]*TimeoutItem, 0, 1024)
```

Comment:
Since it's now one loop per address, the capacity of the array can much smalller.

Reply:
Maybe there are still many request to a store under high concurrency, so we can leave it here.
After all, 8K memory doesn't cost too much.